Remove project usages in internal tasks for better configuration cache support

breskeby opened the pull request:
- This brings us closer to be compliant with the gradle configuration cache feature
   (see https://github.com/elastic/elasticsearch/issues/57918)
- All internal tasks used during precommit should now be free of invalid project references

breskeby committed:
Remove more project usages in tasks for better configuration cache support

- This brings us closer to be compliant with the gradle configuration cache feature
- All internal tasks used during precommit should now be free of invalid project references

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LicenseHeadersTask.groovy:
```diff
@@ -178,4 +180,7 @@ class LicenseHeadersTask extends AntTask {
     }
+
+    @Internal
+    abstract ListProperty<FileCollection> getSourceFolders();
```
Should we be using these "managed" properties more often?

breskeby replied:
I think it makes sense to move our build towards using these

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ForbiddenPatternsTask.java:
```diff
@@ -164,4 +173,11 @@ public void rule(Map<String, String> props) {
+    @Internal
+    abstract ListProperty<FileTree> getSourceFolders();
+
+    void setRootDir(File rootDir) {
```
Why not use a `DirectoryProperty` here?

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitTask.java:
```diff
@@ -31,12 +33,16 @@
     }
 
+    @Inject
+    protected ProjectLayout getProjectLayout() {
```
Where did we land on using method injection vs constructor injection? Should we standardize on one vs the other?

breskeby replied:
I think by default we agreed on ctor injection. In this case I went with method injection due to keeping this local as this is an abstract task and all its descends would need to pass the injected service down here

mark-vieira reviewed (approved):
Some minor comments but otherwise LGTM.

Also, any idea why `TestingConventionsTasks` is plural?

breskeby committed:
Apply review feedback

breskeby commented:
There seems to be quite an performance regression with these changes. I need to dig into this before we can merge that

breskeby committed:
Fix project reference in execution time in licenseheader plugin

breskeby committed:
Fix Path sensitivity for forbidden pattern task

breskeby committed:
Only track path to rootDir not all content

breskeby committed:
Fix imports

breskeby committed:
Fix configuration cache incompatibilities

- configurations are not serializable so we ported it to FileCollection which is enough
for the task properties

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticsearchmachine test this please
